From cfa2a42b83674a9c4c1a71accd82c4ccdac235ae Mon Sep 17 00:00:00 2001
From: Hannu Lyytinen <hannu.lyytinen@unikie.com>
Date: Sat, 31 Jan 2026 09:28:54 +0200
Subject: [PATCH 2/3] firmware: tegra: Add debug prints for BPMP and HSP

Add debug prints to tegra_bpmp_transfer() and related functions
to help diagnose BPMP communication issues when running under
seL4 hypervisor.

Also arm seL4 hypervisor ftrace at the start of tegra_bpmp_probe()
to capture the interesting hypervisor activity during BPMP init.

Signed-off-by: Technology Innovation Institute
Upstream-Status: Inappropriate [TII-specific changes]
---
 drivers/firmware/tegra/bpmp.c | 26 +++++++++++++++++++++++++-
 drivers/mailbox/tegra-hsp.c   |  1 +
 2 files changed, 26 insertions(+), 1 deletion(-)

diff --git a/drivers/firmware/tegra/bpmp.c b/drivers/firmware/tegra/bpmp.c
index 1c01b894ea28..6ba05e166b1a 100644
--- a/drivers/firmware/tegra/bpmp.c
+++ b/drivers/firmware/tegra/bpmp.c
@@ -14,6 +14,7 @@
 #include <linux/pm.h>
 #include <linux/semaphore.h>
 #include <linux/sched/clock.h>
+#include <linux/hyp-ftrace.h>

 #include <soc/tegra/bpmp.h>
 #include <soc/tegra/bpmp-abi.h>
@@ -391,6 +392,8 @@ int tegra_bpmp_transfer_atomic(struct tegra_bpmp *bpmp,
 }
 EXPORT_SYMBOL_GPL(tegra_bpmp_transfer_atomic);

+static int debug_on = 0;
+
 int tegra_bpmp_transfer(struct tegra_bpmp *bpmp,
 			struct tegra_bpmp_message *msg)
 {
@@ -402,22 +405,26 @@ int tegra_bpmp_transfer(struct tegra_bpmp *bpmp,

 	if (!tegra_bpmp_message_valid(msg))
 		return -EINVAL;
-
+	if (debug_on) printk(KERN_ERR "%s:%d\n", __func__, __LINE__);
 	channel = tegra_bpmp_write_threaded(bpmp, msg->mrq, msg->tx.data,
 					    msg->tx.size);
 	if (IS_ERR(channel))
 		return PTR_ERR(channel);

+
+	if (debug_on) printk(KERN_ERR "%s:%d\n", __func__, __LINE__);
 	err = tegra_bpmp_ring_doorbell(bpmp);
 	if (err < 0)
 		return err;

 	timeout = usecs_to_jiffies(bpmp->soc->channels.thread.timeout);

+	if (debug_on) printk(KERN_ERR "%s:%d\n", __func__, __LINE__);
 	err = wait_for_completion_timeout(&channel->completion, timeout);
 	if (err == 0)
 		return -ETIMEDOUT;

+	if (debug_on) printk(KERN_ERR "%s:%d\n", __func__, __LINE__);
 	return tegra_bpmp_channel_read(channel, msg->rx.data, msg->rx.size,
 				       &msg->rx.ret);
 }
@@ -555,7 +562,12 @@ bool tegra_bpmp_mrq_is_supported(struct tegra_bpmp *bpmp, unsigned int mrq)
 	};
 	int err;

+	printk(KERN_ERR "%s:%d entered\n", __func__, __LINE__);
+debug_on=1;
+		printk(KERN_ERR "%s:%d tegra_bpmp_transfer called\n", __func__, __LINE__);
 	err = tegra_bpmp_transfer(bpmp, &msg);
+		printk(KERN_ERR "%s:%d tegra_bpmp_transfer returned\n", __func__, __LINE__);
+		debug_on=0;
 	if (err || msg.rx.ret)
 		return false;

@@ -625,6 +637,8 @@ static int tegra_bpmp_get_firmware_tag_old(struct tegra_bpmp *bpmp, char *tag,
 	void *virt;
 	int err;

+	printk(KERN_ERR "%s:%d entered\n", __func__, __LINE__);
+
 	if (size != TAG_SZ)
 		return -EINVAL;

@@ -656,6 +670,8 @@ static int tegra_bpmp_get_firmware_tag_old(struct tegra_bpmp *bpmp, char *tag,
 static int tegra_bpmp_get_firmware_tag(struct tegra_bpmp *bpmp, char *tag,
 				       size_t size)
 {
+	printk(KERN_ERR "%s:%d entered\n", __func__, __LINE__);
+
 	if (tegra_bpmp_mrq_is_supported(bpmp, MRQ_QUERY_FW_TAG)) {
 		struct mrq_query_fw_tag_response resp;
 		struct tegra_bpmp_message msg = {
@@ -667,10 +683,14 @@ static int tegra_bpmp_get_firmware_tag(struct tegra_bpmp *bpmp, char *tag,
 		};
 		int err;

+	printk(KERN_ERR "%s:%d\n", __func__, __LINE__);
+
 		if (size != sizeof(resp.tag))
 			return -EINVAL;

+		printk(KERN_ERR "%s:%d tegra_bpmp_transfer called\n", __func__, __LINE__);
 		err = tegra_bpmp_transfer(bpmp, &msg);
+		printk(KERN_ERR "%s:%d tegra_bpmp_transfer returned\n", __func__, __LINE__);

 		if (err)
 			return err;
@@ -745,6 +765,10 @@ static int tegra_bpmp_probe(struct platform_device *pdev)
 	size_t size;
 	int err;

 	bpmp = devm_kzalloc(&pdev->dev, sizeof(*bpmp), GFP_KERNEL);
 	if (!bpmp)
 		return -ENOMEM;
@@ -789,6 +813,7 @@ static int tegra_bpmp_probe(struct platform_device *pdev)
 		goto free_mrq;
 	}

+	dev_info(&pdev->dev, "about to get firmware tag\n");
 	err = tegra_bpmp_get_firmware_tag(bpmp, tag, sizeof(tag));
 	if (err < 0) {
 		dev_err(&pdev->dev, "failed to get firmware tag: %d\n", err);
diff --git a/drivers/mailbox/tegra-hsp.c b/drivers/mailbox/tegra-hsp.c
index b781c01a0c5c..62342e1f002b 100644
--- a/drivers/mailbox/tegra-hsp.c
+++ b/drivers/mailbox/tegra-hsp.c
@@ -220,6 +220,7 @@ static irqreturn_t tegra_hsp_doorbell_irq(int irq, void *data)
 		 * It remains to be seen if ignoring the doorbell in that case
 		 * is the correct solution.
 		 */
+		printk(KERN_ERR "%s:%d\n", __func__, __LINE__);
 		if (db && db->channel.chan)
 			mbox_chan_received_data(db->channel.chan, NULL);
 	}
--
2.43.0
